status/status.go

```diff
@@ -148,60 +133,81 @@ OuterLoop:
-					} else {
-						log.Printf("Mattermost updated")
-					}
-				}
 			}
 			first = false
 			time.Sleep(time.Second)
 		}
 	}
 }
 
-func UpdateMattermost(nc *http.Client, status bool) error {
-	mm := model.NewAPIv4Client(configuration.MattermostServer)
-	mm.HttpClient = nc
-	mm.SetToken(configuration.MattermostToken)
-
-	channel, resp := mm.GetChannel(configuration.MattermostChannelId, "")
-	if channel == nil {
-		return fmt.Errorf("Get channel: %+v", resp)
+// UpdateTopic modifies the topic (IRC, Mattermost) by matching `re` and replacing
+// the subexpression by `new`. The regexp must have exactly one subexpression.
+func (ss *SWITCHSTATE) UpdateTopic(irccon *irc.Connection, nc *http.Client, re *regexp.Regexp, new string) {
+	err := ss.updateTopicIRC(irccon, re, new)
+	if err != nil {
+		log.Printf("updateTopicIRC error: %s\n", err)
 	}
 
-	const labOpen = "|| LAB OPEN ||"
-	const labClosed = "|| LAB CLOSED ||"
-
-	var strStatus string
-	if status {
-		strStatus = labOpen
-	} else {
-		strStatus = labClosed
+	if configuration.MattermostServer != "" {
+		err = ss.updateTopicMattermost(nc, re, new)
+		if err != nil {
+			log.Printf("updateTopicMattermost error: %s\n", err)
+		}
 	}
+}
 
-	var newHeader string
-	if strings.Contains(channel.Header, labOpen) {
-		newHeader = strings.ReplaceAll(channel.Header, labOpen, strStatus)
-	} else if strings.Contains(channel.Header, labClosed) {
-		newHeader = strings.ReplaceAll(channel.Header, labClosed, strStatus)
+func (ss *SWITCHSTATE) updateTopicIRC(irccon *irc.Connection, re *regexp.Regexp, new string) error {
+	match := re.FindStringSubmatchIndex(ss.Topic)
+	if len(match) == 4 {
+		start, end := match[2], match[3]
+		topic := ss.Topic[:start] + new + ss.Topic[end:]
+		if ss.Topic != topic {
+			log.Printf("New IRC topic: %q\n", topic)
+			if configuration.TopicUseChanserv {
+				irccon.Privmsg("ChanServ", fmt.Sprintf("TOPIC %s %s", configuration.BotChannel, topic))
+			} else {
+				irccon.SendRawf("TOPIC %s :%s", BotChannel, topic)
+			}
+			ss.Topic = topic
+		} else {
+			log.Printf("IRC topic unchanged")
+		}
 	} else {
-		return fmt.Errorf("Channel header didn't have the key phrase: %q", channel.Header)
+		return fmt.Errorf("IRC topic %q did not match regexp %q", ss.Topic, re)
 	}
+	return nil
+}
 
-	log.Printf("New Mattermost header: %q", newHeader)
+func (ss *SWITCHSTATE) updateTopicMattermost(nc *http.Client, re *regexp.Regexp, new string) error {
+	mm := model.NewAPIv4Client(configuration.MattermostServer)
+	mm.HttpClient = nc
+	mm.SetToken(configuration.MattermostToken)
 
-	updated, resp := mm.PatchChannel(channel.Id, &model.ChannelPatch{
-		Header: &newHeader,
-	})
-	if updated == nil {
-		return fmt.Errorf("Update channel: %+v", resp)
+	channel, resp := mm.GetChannel(configuration.MattermostChannelId, "")
+	if channel == nil {
+		log.Printf("Mattermost error: Get channel: %+v\n", resp)
+	} else {
+		match := re.FindStringSubmatchIndex(channel.Header)
+		if len(match) == 4 {
```

Comment:
this is not actually a comment, but man this regex API actually made me cry a bit